compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenDialect.cpp

```diff
@@ -52,41 +52,177 @@ void IREECodegenDialect::initialize() {
+  // Ensure the operation has exactly one argument of type any_op.
+  if (argTypes.size() != 1 || argTypes.front() != anyOpType) {
+    return foreachMatchOp.emitError()
+           << "'ForeachMatchOp' must take exactly one 'any_op' argument"
+           << kRequiredByDefaultAttr;
+  }
+
+  SmallVector<Type> resultTypes(foreachMatchOp.getResultTypes());
+  // Ensure the operation has exactly one result of type any_op.
+  if (resultTypes.size() != 1 || resultTypes.front() != anyOpType) {
+    return foreachMatchOp.emitError()
+           << "'ForeachMatchOp' must return exactly one 'any_op' result"
+           << kRequiredByDefaultAttr;
+  }
+
+  return success();
+}
+
 LogicalResult
 IREECodegenDialect::verifyOperationAttribute(Operation *op,
                                              NamedAttribute attribute) {
   StringRef symbol = attribute.getName().strref();
   Attribute attr = attribute.getValue();
   // This function verifies the validity of a specific operation attribute.
-  // - If the attribute's name matches `kTuningDefaultSpecAttrName`, make
-  //   sure it contains a single named sequence op with name `__kernel_config`.
+  // - If the attribute's name matches kTuningSpecDefaultEntrypointAttrName
+  // (`iree_codegen.tuning_spec_with_default_entrypoint`):
+  //   1. Ensure that the module contains a single named sequence operation with
+  //   the name `__kernel_config`.
+  //   2. Verify that this `__kernel_config` named sequence operation has the
+  //   attribute `iree_codegen.tuning_spec_entrypoint`.
+  //   3. Ensure that the named sequence operation contains exactly **one**
+  //   `ForeachMatchOp`.
+  //      - ForeachMatchOp must not have `flatten_results` and `restrict_root`
+  //        attributes.
+  //      - ForeachMatchOp must have exactly one argument of type any_op.
+  //      - ForeachMatchOp must have exactly one result of type any_op.
+  //   4. Ensure that only one named sequence operation with the
+  //   `iree_codegen.tuning_spec_entrypoint` attribute.
   // - If the attribute's name matches `kTuningSpecEntrypointAttrName`
-  // ("iree_codegen.tuning_spec_entrypoint"):
+  // (`iree_codegen.tuning_spec_entrypoint`):
   //   1. The attribute value must be a UnitAttr.
   //   2. If the operation is a transform::NamedSequenceOp:
   //      - The operation's function signature must satisfy the following:
-  //         a. It must have exactly one result type, and the result must be of
-  //         type `transform::AnyOpType`.
+  //         a. It must have exactly one result type, and the result must be
+  //         of type `transform::AnyOpType`.
   //         b. It must have exactly one argument type, and the argument must be
   //         of type `transform::AnyOpType`.
 
   if (symbol == kTuningSpecDefaultEntrypointAttrName) {
+    const std::string kRequiredByDefaultAttr =
+        " (required by '" + std::string(kTuningSpecDefaultEntrypointAttrName) +
+        "').";
```

Comment:
The variable name is misleading in two ways: it's not a compile time constant, so we shouldn't use the `k` prefix, and this is a string instead of an attribute. I'd call it something like `requiredByDefaultAttrMessage` or `requiredByMessage` or `requiredByErrorSuffix`.

You can also construct this with `llvm::formatv`, but either way is fine IMO.

Reply:
Also, we typically don't end error messages with `.` in mlir